core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -0,0 +1,613 @@
+
+class half_t {
+ public:
+  using impl_type = HALF_IMPL_TYPE;
+
+ private:
+  impl_type val;
+
+ public:
+  // Conversion operator for __half(bar) = half_t(foo)
+  KOKKOS_FUNCTION
+  operator impl_type() const { return val; }
+
+  KOKKOS_FUNCTION
+  explicit operator bool() const {
+    return static_cast<bool>(__half2float(val) == 0.0F);
+  }
+
+  KOKKOS_FUNCTION
+  half_t(const half_t&) = default;
+
+  KOKKOS_FUNCTION
+  half_t(impl_type rhs = cast_to_half(0)) : val(rhs) {}
+
+  // Cast rhs to half for assignment to lhs of type half_t
+  template <class T>
+  KOKKOS_FUNCTION half_t(T rhs) : half_t(cast_to_half(rhs)) {}
+
+  // Unary operators
+  KOKKOS_FUNCTION
+  half_t operator+() const {
+    half_t tmp = *this;
+#ifdef __CUDA_ARCH__
+    tmp.val = +tmp.val;
+#else
+    tmp.val   = __float2half(+__half2float(tmp.val));
+#endif
+    return tmp;
+  }
+
+  KOKKOS_FUNCTION
+  half_t operator-() const {
+    half_t tmp = *this;
+#ifdef __CUDA_ARCH__
+    tmp.val = -tmp.val;
+#else
+    tmp.val   = __float2half(-__half2float(tmp.val));
+#endif
+    return tmp;
+  }
+
+  // Prefix operators
+  KOKKOS_FUNCTION
+  half_t& operator++() {
+#ifdef __CUDA_ARCH__
+    ++val;
+#else
+    float tmp = __half2float(val);
+    ++tmp;
+    val       = __float2half(tmp);
+#endif
+    return *this;
+  }
+
+  KOKKOS_FUNCTION
+  half_t& operator--() {
+#ifdef __CUDA_ARCH__
+    --val;
+#else
+    float tmp = __half2float(val);
+    --tmp;
+    val     = __float2half(tmp);
+#endif
+    return *this;
+  }
+
+  // Postfix operators
+  KOKKOS_FUNCTION
+  half_t operator++(int) {
+    half_t tmp = *this;
+    operator++();
+    return tmp;
+  }
+
+  KOKKOS_FUNCTION
+  half_t operator--(int) {
+    half_t tmp = *this;
+    operator--();
+    return tmp;
+  }
+
+  // Binary operators
+  KOKKOS_FUNCTION
+  half_t& operator=(impl_type rhs) {
+    val = rhs;
+    return *this;
+  }
+
+  template <class T>
+  KOKKOS_FUNCTION half_t& operator=(T rhs) {
+    val = cast_to_half(rhs).val;
+    return *this;
+  }
+
+  // Compound operators
+  KOKKOS_FUNCTION
+  half_t& operator+=(half_t rhs) {
+#ifdef __CUDA_ARCH__
+    val += rhs.val;
+#else
+    val     = __float2half(__half2float(val) + __half2float(rhs.val));
+#endif
+    return *this;
+  }
+
+  KOKKOS_FUNCTION
+  half_t& operator-=(half_t rhs) {
+#ifdef __CUDA_ARCH__
+    val -= rhs.val;
+#else
+    val     = __float2half(__half2float(val) - __half2float(rhs.val));
+#endif
+    return *this;
+  }
+
+  KOKKOS_FUNCTION
+  half_t& operator*=(half_t rhs) {
+#ifdef __CUDA_ARCH__
+    val *= rhs.val;
+#else
+    val     = __float2half(__half2float(val) * __half2float(rhs.val));
+#endif
+    return *this;
+  }
+
+  KOKKOS_FUNCTION
+  half_t& operator/=(half_t rhs) {
+#ifdef __CUDA_ARCH__
+    val /= rhs.val;
+#else
+    val     = __float2half(__half2float(val) / __half2float(rhs.val));
+#endif
+    return *this;
+  }
+
+  // Binary Arithmetic
+  KOKKOS_FUNCTION
+  half_t operator+(half_t rhs) const {
```

Comment:
These binary operators should be hidden friends, as in:

```
half_t friend operator+(half_t lhs, half_t rhs)
{
 #ifdef __CUDA_ARCH__
     lhs.val += rhs.val;
 #else
     lhs.val = __float2half(__half2float(lhs.val) + __half2float(rhs.val));
 #endif
     return lhs;
}
```

The benefits are symmetry and ADL.

It is less common to do those for unary operators, although Anthony Williams makes a case for it at [https://www.justsoftwaresolutions.co.uk/cplusplus/hidden-friends.html](https://www.justsoftwaresolutions.co.uk/cplusplus/hidden-friends.html).